feat: add dms rocketmq instance (#2556)
huaweicloud/terraform-provider-huaweicloud: docs/resources/dms_rocketmq_instance.md

---
subcategory: "Distributed Message Service (DMS)"
---

# huaweicloud_dms_rocketmq_instance

>   Manage DMS RocketMQ instance resources within HuaweiCloud.

Reviewer: ```suggestion
Manages DMS RocketMQ instance resources within HuaweiCloud.
```

Author: done

  By default, public access is disabled.
  Changing this parameter will create a new resource.

* `publicip_id` - (Optional, String, ForceNew) Specifies the ID of the EIP bound to the instance.
  Use commas (,) to separate multiple EIP IDs.
  This parameter is mandatory if public access is enabled (that is, enable_publicip is set to true).
  Changing this parameter will create a new resource.

* `broker_num` - (Optional, Int, ForceNew) Specifies the broker numbers. Defaults to 1.
  Changing this parameter will create a new resource.

* `retention_policy` - (Optional, Bool) Specifies the ACL access control.

## Attributes Reference

In addition to all arguments above, the following attributes are exported:

* `id` - Specifies a resource ID in UUID format.

* `status` - Indicates the status of the DMS RocketMQ instance.

* `type` - Indicates the DMS RocketMQ instance type. Value: cluster.

* `specification` - Indicates the instance specification. For a cluster DMS RocketMQ instance, VM specifications
  and the number of nodes are returned.

* `subnet_cidr` - Indicates the subnet route address.

* `user_id` - Indicates the ID of the user who created the DMS RocketMQ instance.

> * `user_name` - Indicates the name of the user who created the DMS RocketMQ instance.

Reviewer: I think the two attributes are meaningless

Author: done

---
subcategory: "Distributed Message Service (DMS)"
---

# huaweicloud_dms_rocketmq_instance

  Manage DMS RocketMQ instance resources within HuaweiCloud.

## Example Usage

```HCL
resource "huaweicloud_dms_rocketmq_instance" "test" {
  name                = "rocketmq_name_test"
  description         = "this is a rocketmq instance"
  engine              = "reliability"
  engine_version      = "4.8.0"
  total_storage_space = 10
  vpc_id              = "4579f2f5-cbe8-425a-8f32-53dcb9d9053a"
  subnet_id           = "4579f2f5-cbe8-425a-8f32-53dcb9d9053a"
  security_group_id   = "4579f2f5-cbe8-425a-8f32-53dcb9d9053a"
  available_zones     = ["cn-north-4a", "cn-north-4b"]
  product_id          = "c6.4u8g.cluster"
  storage_spec_code   = "dms.physical.storage.high.v2"

> }

Reviewer 2: ```suggestion
variable "vpc_id" {}
variable "subnet_id" {}
variable "security_group_id" {}
variable "availability_zones" {
  type = list(string)
}

resource "huaweicloud_dms_rocketmq_instance" "test" {
  name                = "rocketmq_name_test"
  description         = "this is a rocketmq instance"
  engine              = "reliability"
  engine_version      = "4.8.0"
  total_storage_space = 10
  vpc_id              = var.vpc_id
  subnet_id           = var.subnet_id
  security_group_id   = var.security_group_id
  availability_zones  = var.availability_zones
  product_id          = "c6.4u8g.cluster"
  storage_spec_code   = "dms.physical.storage.high.v2"
}
```

Author: done

```

## Argument Reference

The following arguments are supported:

* `region` - (Optional, String, ForceNew) Specifies the region in which to create the resource.
  If omitted, the provider-level region will be used. Changing this parameter will create a new resource.

* `name` - (Required, String) Instance name.
  An instance name starts with a letter, consists of 4 to 64 characters, and can contain only letters,
  digits, underscores (_), and hyphens (-).

* `engine` - (Required, String, ForceNew) Message engine. Value: reliability.
  Changing this parameter will create a new resource.

* `engine_version` - (Required, String, ForceNew) Version of the message engine. Value: 4.8.0.
  Changing this parameter will create a new resource.

> * `total_storage_space` - (Required, Int, ForceNew) Total storage space.

Reviewer 3: add the value range description as there is limit in the code

Author: done

  Changing this parameter will create a new resource.

* `vpc_id` - (Required, String, ForceNew) VPC ID.

  Changing this parameter will create a new resource.

* `subnet_id` - (Required, String, ForceNew) Subnet information.
  Changing this parameter will create a new resource.

* `security_group_id` - (Required, String) ID of the security group to which the instance belongs.

> * `availability_zone` - (Required, List, ForceNew) Name of the AZ where instance brokers reside and which
>   has available resources.

Reviewer 2: ```suggestion
* `availability_zones` - (Required, List, ForceNew) The list of availability zone names, where instance brokers reside and which
  has available resources.
```

Author: done

  Changing this parameter will create a new resource.

> * `product_id` - (Required, String, ForceNew) RocketMQ instance specifications. The options are as follows:

Reviewer 2: `flavor_id` please

Author: done

  + **c6.4u8g.cluster**: maximum number of topics on each broker: 4000; maximum number of consumer groups
    on each broker: 4000
  + **c6.8u16g.cluster**: maximum number of topics on each broker: 8000; maximum number of consumer groups
    on each broker: 8000
  + **c6.12u24g.cluster**: maximum number of topics on each broker: 12,000; maximum number of consumer groups
    on each broker: 12,000
  + **c6.16u32g.cluster**: maximum number of topics on each broker: 16,000; maximum number of consumer groups
    on each broker: 16,000
  Changing this parameter will create a new resource.

* `storage_spec_code` - (Required, String, ForceNew) Storage I/O specification. The options are as follows:
  + **dms.physical.storage.high.v2**: high I/O disk
  + **dms.physical.storage.ultra.v2**: ultra-high I/O disk
  Changing this parameter will create a new resource.

* `description` - (Optional, String) Instance description.
  The description can contain a maximum of 1024 characters.

> * `ssl_enable` - (Optional, Bool, ForceNew) Whether to enable SSL encryption for access.

Reviewer 3: add default value description here, for example 'Defaults to false'

Author: done

  Changing this parameter will create a new resource.

> * `ipv6_enable` - (Optional, Bool, ForceNew) Whether to support IPv6.

Reviewer 2: ditto.

Author: done

  Changing this parameter will create a new resource.

* `enable_publicip` - (Optional, Bool, ForceNew) Whether to enable public access.
  By default, public access is disabled.
  Changing this parameter will create a new resource.

* `publicip_id` - (Optional, String, ForceNew) ID of the EIP bound to the instance.
  Use commas (,) to separate multiple EIP IDs.
  This parameter is mandatory if public access is enabled (that is, enable_publicip is set to true).
  Changing this parameter will create a new resource.

> * `broker_num` - (Optional, Int, ForceNew) Number of brokers.

Reviewer 3: add default value description here

Author: done

  Changing this parameter will create a new resource.

> * `retention_policy` - (Optional, Bool) ACL.

Reviewer 2: The description is too short...
and please check other parameters

Author: done

## Attributes Reference

In addition to all arguments above, the following attributes are exported:

* `id` - The resource ID.

* `status` - MStatus.

* `type` - Instance type. Value: cluster.

* `specification` - Instance specification.

> * `instance_id` - Instance ID.

Reviewer 2: What's the different between `id` and `instance_id`?

Author: the param instance_id has been deleted

> * `charging_mode` - Billing mode. 1 indicates pay-per-use.

Reviewer 2: useless attribute.
And this value is not `1`, it should be `postPaid`.

Author: done

> * `vpc_name` - VPC name.

Reviewer 2: useless attribute.
and please check other parameters.

Author: done

* `maintain_end` - Time at which the maintenance window ends. The format is HH:mm:ss.

* `enable_log_collection` - Whether log collection is enabled.

* `used_storage_space` - Used message storage space. Unit: GB.

* `publicip_address` - Public IP address.

* `cross_vpc_info` - Cross-VPC access information.

* `storage_resource_id` - Storage resource ID.

* `service_type` - Service type.

* `storage_type` - Storage category.

* `extend_times` - Extension time.

* `support_features` - Features supported by the instance.

* `disk_encrypted` - Whether disk encryption is enabled.

* `ces_version` - Cloud Eye version.

* `node_num` - Node quantity.

* `new_spec_billing_enable` - Whether billing based on new specifications is enabled.

* `enable_acl` - Whether access control is enabled.

* `namesrv_address` - Metadata address.

* `broker_address` - Service data address.

* `public_namesrv_address` - Public network metadata address.

* `public_broker_address` - Public network service data address.

* `total_storage_space` - Total storage space.

* `resource_spec_code` - Resource specifications.

## Import

The rocketmq instance can be imported using the `id`, e.g.

```

> $ terraform import huaweicloud_dms_rocketmq_instance.test 0ce123456a00f2591fabc00385ff1234

Reviewer 2: actually, the ID format of RockMQ instance is `xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx`.

Author: done
